drivers/usb/device/usb_dc_sam_usbc.c

```diff
@@ -0,0 +1,1160 @@
+static struct usb_device_data dev_data;
+static volatile Usbc *regs = (Usbc *) DT_INST_REG_ADDR(0);
+static uint32_t num_pins = ATMEL_SAM_DT_INST_NUM_PINS(0);
+static struct soc_gpio_pin pins[] = ATMEL_SAM_DT_INST_PINS(0);
+
+#if defined(CONFIG_USB_DRIVER_LOG_LEVEL_DBG)
+static uint32_t dev_ep_sta_dbg[2][NUM_OF_EP_MAX];
+
+static void usb_dc_sam_usbc_clean_sta_dbg(void)
+{
+	int i;
+
+	for (i = 0; i < NUM_OF_EP_MAX; i++) {
+		dev_ep_sta_dbg[0][i] = 0;
+		dev_ep_sta_dbg[1][i] = 0;
+	}
+}
+#else
+#define usb_dc_sam_usbc_clean_sta_dbg()
+#endif
+
+static int usb_dc_sam_usbc_ep_curr_bank(int ep_idx)
+{
+	int idx = ep_idx * 2;
+
+	if ((ep_idx > 0) &&
+	    (regs->UESTA[ep_idx] & USBC_UESTA0_CURRBK(1)) > 0) {
+		idx++;
+	}
+
+	return idx;
+}
+
+static bool usb_dc_is_attached(void)
+{
+	return (regs->UDCON & USBC_UDCON_DETACH) == 0;
+}
+
+static bool usb_dc_ep_is_enabled(uint8_t ep_idx)
+{
+	int reg = regs->UERST;
+
+	return (reg & BIT(USBC_UERST_EPEN0_Pos + ep_idx));
+}
+
+static int usb_dc_sam_usbc_ep_alloc_buf(int ep_idx)
+{
+	struct sam_usbc_desc_table_t *ep_desc_bk;
+	bool ep_enabled[NUM_OF_EP_MAX];
+	int desc_mem_alloc;
+	int mps;
+
+	if (ep_idx >= NUM_OF_EP_MAX) {
+		return -EINVAL;
+	}
+
+	desc_mem_alloc = 0;
+
+	mps = dev_data.ep_data[ep_idx].mps_x2
+	    ? dev_data.ep_data[ep_idx].mps * 2
+	    : dev_data.ep_data[ep_idx].mps;
+
+	/* Check if there are memory to all endpoints */
+	for (int i = 0; i < NUM_OF_EP_MAX; i++) {
+		if (!dev_data.ep_data[i].is_configured || i == ep_idx) {
+			continue;
+		}
+
+		desc_mem_alloc += dev_data.ep_data[i].mps_x2
+				? dev_data.ep_data[i].mps * 2
+				: dev_data.ep_data[i].mps;
+	}
+
+	if ((desc_mem_alloc + mps) > USBC_RAM_SIZE) {
+		memset(&dev_data.ep_data[ep_idx], 0,
+		       sizeof(struct usb_device_ep_data));
+
+		return -ENOMEM;
+	}
+
+	/*
+	 * Allocate the memory. This part is a bit tricky as memory can only
+	 * be allocated if all above endpoints are disabled and not allocated.
+	 * Loop backward through the above endpoints, disable them if they are
+	 * enabled, deallocate their memory if needed. Then loop again through
+	 * all the above endpoints to allocate and enabled them.
+	 */
+	for (int i = NUM_OF_EP_MAX - 1; i >= ep_idx; i--) {
+		ep_enabled[i] = usb_dc_ep_is_enabled(i);
+
+		if (ep_enabled[i]) {
+			LOG_DBG("Temporary disable ep idx 0x%02x", i);
+			usb_dc_ep_disable(i);
+		}
+	}
+
+	desc_mem_alloc = 0U;
+	for (int i = 0; i < ep_idx; i++) {
+		if (!dev_data.ep_data[i].is_configured) {
+			continue;
+		}
+
+		desc_mem_alloc += dev_data.ep_data[i].mps_x2
+				? dev_data.ep_data[i].mps * 2
+				: dev_data.ep_data[i].mps;
+	}
+
+	ep_desc_bk = ((struct sam_usbc_desc_table_t *) &dev_desc)
+		   + (ep_idx * 2);
+	for (int i = ep_idx; i < NUM_OF_EP_MAX; i++) {
+		if (!dev_data.ep_data[i].is_configured && (i != ep_idx)) {
+			ep_desc_bk += 2;
+			continue;
+		}
+
+		/* Alloc bank 0 */
+		ep_desc_bk->ep_pipe_addr = ((uint8_t *) USBC_RAM_ADDR)
```

Comment:
Is that dedicated USB device controller RAM?

Reply:
No, USB controller have their own dedicated DMA and can access that memory. That was dedicated to USB stack on Zephyr.

https://github.com/zephyrproject-rtos/zephyr/blob/f5a3deaad382b0cb522cbfd516a9b44eb7ca3311/dts/arm/atmel/sam4l.dtsi#L42-L51